### PR TITLE
Faster update container

### LINKED ISSUE #26: LinkedCells::updateContainer() can duplicate or remove particles if halo particles are present
LinkedCells::updateContainer() should only be called if all halo particles are DELETED.
There should be a run-time check for that, as it is not to expensive to check (still on the high-level).

problem:
if we don't do that particles can be duplicated, or lost, as they can no longer be differentiated:
```
addParticle(p1);
addHaloParticle(p2);
p1->setR(); // set the position of the normal particle to somewhere outside of the domain
updateContainer();
// now both particle and haloParticle are outside of the domain
```

possible implementation:
* bool that indicates whether halos are present
* if haloparticle is added, this bool is set to true
* if all haloparticles are deleted, this bool is set to false

other possible solution (more complicated!):
* don't store halo and leaving particles together

### LINKED ISSUE #72: updateContainer
**Is your feature request related to a problem? Please describe.**
`LinkedCells::updateContainer()` currently deletes all particles and adds them again.

**Describe the solution you'd like**
Do not delete particles, only move those which crossed cell borders.

**Describe alternatives you've considered**
Use Functors?!

### PR DESCRIPTION
**Pull request by FG-TUM**
Faster and more parallel implementation of `LinkedCells::updateContainer()`

- [x] locking mechanism for `FullParticleCell::addParticle()`
- [x] full parallel implementation of `LinkedCells::updateContainer()`
- [x] unit tests for `LinkedCells::updateContainer()`
- [x] resolves #72 
- [x] resolves #26  

Tested on my machine (i7-7700 8x3.6GHz)
- 1,000,000 Particles (uniform random distribution)
- 52x 52 x 52 Cells
- random = particle jumps to a complete random location in the domain
- realistic = particle jumps to a random neighbor cell
- omitting locks (= allowing data races) has no significant impact

Particle relocations [ms] |  no | random 50%  | random 100% | realistic 5% | realistic 10%
------------ | ------------- | ------------- | ------------- | ------------- | -------------
old                         | 180 | 285 | 285 | 185 | 190
lock everything      |   30 | 240 | 250 |   42 | 50
add-barrier-delete |   35 | 250 | 288 |   47 | 58

Versions can be compared in 2330c573c0dde6b5da3b1a1eec670edf915ccf0f

### CODE REVIEW

File: src/autopas/containers/LinkedCells.h [hunk 1]
```diff
@@ -112,31 +112,33 @@ class LinkedCells : public ParticleContainer<Particle, ParticleCell, SoAArraysTy
...
-    for (auto &particle : invalidParticles) {
-      if (inBox(particle.getR(), this->getBoxMin(), this->getBoxMax())) {
-        addParticle(particle);
-      } else {
-        addHaloParticle(particle);
+    for (size_t cellId = 0; cellId < this->getCells().size(); ++cellId) {
+      // if empty
+      if (not this->getCells()[cellId].isNotEmpty()) continue;
+
+      std::array<double, 3> cellLoweCorner, cellUpperCorner;
```
Comment (SteffenSeckler) [inline src/autopas/containers/LinkedCells.h]: LoweR

File: src/autopas/containers/LinkedCells.h [hunk 2]
```diff
@@ -112,31 +112,33 @@ class LinkedCells : public ParticleContainer<Particle, ParticleCell, SoAArraysTy
...
-//#pragma omp parallel reduction(vecMerge: invalidParticles)
-#ifdef AUTOPAS_OPENMP
-#pragma omp parallel
-#endif  // AUTOPAS_OPENMP
-    {
-      std::vector<Particle> myInvalidParticles;
-      for (auto iter = this->begin(); iter.isValid(); ++iter) {
-        myInvalidParticles.push_back(*iter);
-      }
+    auto haloIter = this->getRegionIterator(this->_cellBlock.getHaloBoxMin(), this->_cellBlock.getHaloBoxMax(),
```
Comment (SteffenSeckler) [inline src/autopas/containers/LinkedCells.h]: should be done with the normal iterator (+ haloCellsOnly) and not with the RegionIterator.
you will anyways iterate through all halo Cells, which is exactly what the normal iterator will do.

File: src/autopas/containers/LinkedCells.h [hunk 3]
```diff
@@ -112,31 +112,33 @@ class LinkedCells : public ParticleContainer<Particle, ParticleCell, SoAArraysTy
...
+
+      for (auto &&pIter = this->getCells()[cellId].begin(); pIter.isValid(); ++pIter) {
+        // if not in cell
+        if (notInBox(pIter->getR(), cellLoweCorner, cellUpperCorner)) {
+          // if not in halo
+          if (inBox(pIter->getR(), this->getBoxMin(), this->getBoxMax()))
+            addParticle(*pIter);
+          else
+            addHaloParticle(*pIter);
+          pIter.deleteCurrentParticle();
```
Comment (SteffenSeckler) [inline src/autopas/containers/LinkedCells.h]: nope :P
deleteCurrentParticle, pIter.isValid() (needs Cell::getNumParticles() ) etc. are not thread safe.

Comment (SteffenSeckler) [reply src/autopas/containers/LinkedCells.h]: alternatives:
1. make everything thread safe
2. safe to temporary vector -> add afterwards (with or without parallelization)

Comment (FG-TUM) [reply src/autopas/containers/LinkedCells.h]: both alternatives implemented and compared. See text at the top.

File: src/autopas/containers/LinkedCells.h (hunk 2 above)
Comment (FG-TUM) [reply src/autopas/containers/LinkedCells.h]: fixed 1084179

File: src/autopas/containers/LinkedCells.h (hunk 1 above)
Comment (FG-TUM) [reply src/autopas/containers/LinkedCells.h]: fixed d7c84c0f2d066bc59baaa1a2f87ec78a83c140bb

**Pull request merged**